zk/datastream/server/data_stream_server.go

```diff
@@ -206,78 +205,95 @@ func (srv *DataStreamServer) CreateStreamEntriesProto(
+	if isBatchEnd {
 		entryCount++
 	}
 
 	entries := make([]DataStreamEntryProto, entryCount)
 	index := 0
 
 	// BATCH BOOKMARK
-	if batchNumber != lastBatchNumber {
-		for i := 0; i < int(batchNumber-lastBatchNumber); i++ {
-			workingBatch := lastBatchNumber + uint64(i)
-			nextWorkingBatch := workingBatch + 1
-
-			// handle any gers that need to be written before closing the batch down
-			for _, ger := range gers {
-				if ger.BatchNumber == workingBatch {
-					entries[index] = &ger
-					index++
+	if batchStart {
+		// if we have a gap of more than 1 batch then we need to write in the batch start and ends for these empty batches
+		if batchGap > 1 {
+			for i := 1; i < int(batchGap); i++ {
+				workingBatch := lastBatchNumber + uint64(i)
+
+				// bookmark for new batch
+				batchBookmark := srv.CreateBatchBookmarkEntryProto(workingBatch)
+				entries[index] = batchBookmark
+				index++
+
+				// new batch starting
+				batchType := datastream.BatchType_BATCH_TYPE_REGULAR
+				if batchNumber == 1 {
+					batchType = datastream.BatchType_BATCH_TYPE_INJECTED
+				}
+				fork, err := reader.GetForkId(workingBatch)
+				if err != nil {
+					return nil, err
+				}
+				batch := srv.CreateBatchStartProto(workingBatch, srv.chainId, fork, batchType)
+				entries[index] = batch
+				index++
+
+				// see if we have any gers to handle
+				for _, ger := range gers {
+					upd := ger.UpdateGER
+					if upd.BatchNumber == workingBatch {
+						gerUpdate := srv.CreateGerUpdateProto(upd.BatchNumber, upd.Timestamp, libcommon.BytesToHash(upd.GlobalExitRoot), libcommon.BytesToAddress(upd.Coinbase), upd.ForkId, upd.ChainId, libcommon.BytesToHash(upd.StateRoot))
+						entries[index] = gerUpdate
+						index++
+					}
 				}
-			}
 
-			// we only write the batch end if we are in the standard stream mode - executors don't start the stream
-			// bytes with a batch end.  Instead, we will write this at the end of the stream
-			if mode != ExecutorOperationMode {
+				// seal off the last batch
 				localExitRoot, err := srv.getLocalExitRoot(workingBatch, reader, tx)
 				if err != nil {
 					return nil, err
 				}
-				// seal off the last batch
 				root := lastBlock.Root()
 				end := srv.CreateBatchEndProto(localExitRoot, root, workingBatch)
 				entries[index] = end
 				index++
 			}
+		}
 
-			// bookmark for new batch
-			batchBookmark := srv.CreateBatchBookmarkEntryProto(nextWorkingBatch)
-			entries[index] = batchBookmark
-			index++
-
-			// new batch starting
-			batchType := datastream.BatchType_BATCH_TYPE_REGULAR
-			if batchNumber == 1 {
-				batchType = datastream.BatchType_BATCH_TYPE_INJECTED
-			}
-			fork, err := reader.GetForkId(nextWorkingBatch)
-			if err != nil {
-				return nil, err
-			}
-			batch := srv.CreateBatchStartProto(nextWorkingBatch, srv.chainId, fork, batchType)
-			entries[index] = batch
-			index++
+		// now write in the batch start for this batch
+		batchType := datastream.BatchType_BATCH_TYPE_REGULAR
```

Comment:
Looks like this duplicates with what is within the `for` loop — I would extract a method or a function for that.

Reply:
I'm introducing a type here to track adding entries to the collection to reduce the chance of bugs in the future as well and to have a type we can pass around to functions for this kind of thing